README.md
<br>

<div align="center">
<img width="456" src="https://raw.githubusercontent.com/wayofdev/docker-project-services/master/assets/logo.gh-light-mode-only.png#gh-light-mode-only" alt="Logo for light mode">
<img width="456" src="https://raw.githubusercontent.com/wayofdev/docker-project-services/master/assets/logo.gh-dark-mode-only.png#gh-dark-mode-only" alt="Logo for dark mode">
</div>
<br>
<br>

<div align="center">
<a href="https://actions-badge.atrox.dev/wayofdev/docker-project-services/goto"><img alt="Build Status" src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fwayofdev%2Fdocker-project-services%2Fbadge&style=flat-square"/></a>
<a href="https://github.com/wayofdev/docker-project-services/tags"><img src="https://img.shields.io/github/v/tag/wayofdev/docker-project-services?sort=semver&style=flat-square" alt="Latest Version"></a>
<a href="LICENSE.md"><img src="https://img.shields.io/github/license/wayofdev/docker-project-services.svg?style=flat-square&color=blue" alt="Software License"/></a>
<a href="#"><img alt="Commits since latest release" src="https://img.shields.io/github/commits-since/wayofdev/docker-project-services/latest?style=flat-square"></a>
</div>


<br>

# Docker Project Services

This Repository is used together with SOA based projects, consisting of microservices, built on [laravel-starter-tpl](https://github.com/wayofdev/laravel-starter-tpl).

### → Purpose

The purpose of this repository is to contain Docker services that can be shared across all microservices in one domain.

<br>

## 🚩 Requirements

* **macOS** Monterey or **Linux**
* **Docker** 20.10 or newer
  - [How To Install and Use Docker on Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-22-04)
* **Cloned, configured and running** [docker-shared-services](https://github.com/wayofdev/docker-shared-services), to support system-wide DNS, Routing and SSL support via Traefik.

<br>

## 💻 Usage

> **Warning**: Repository with [docker-shared-services](https://github.com/wayofdev/docker-shared-services) should be configured, up and running before

### → Instructions

1. Clone repository:

   ```bash
   $ git clone git@github.com:wayofdev/docker-project-services.git
   ```

2. Generate default .env file:

   ```bash
   # With custom namespace provided, it will be used to prefix all services
   # in Docker network for current project
   $ make env COMPOSE_PROJECT_NAME=bfq

   # Default project namespace will be used - **wod**
   $ make env
   ```

   Edit the created .env file if necessary. You may want to change the default domain.

3. (Optional) Enable the `docker-compose.override.yaml` file to run extra services, like pg-admin and others:
   ```bash
   $ make override
   ```

4. Run this repository:

   ```bash
   $ make up
   ```

5. Check that everything works:

   ```bash
   $ make ps
   $ make logs
   ```

<br>

## 🧪 Testing

You can check `Makefile` to get full list of commands for local testing. For testing, you can use these commands to test whole role or separate tasks:

Testing docker-compose using dcgoss:

```bash
$ make test
```

<br>

## 🤝 License

[![Licence](https://img.shields.io/github/license/wayofdev/docker-project-services?style=for-the-badge&color=blue)](./LICENSE)

<br>

## 🙆🏼‍♂️ Author Information

This repository was created in **2022** by [lotyp / wayofdev](https://github.com/wayofdev).

<br>

## 🫡 Contributors

<img align="left" src="https://img.shields.io/github/contributors-anon/wayofdev/docker-project-services?style=for-the-badge" alt="Contributors"/>

<a href="https://github.com/wayofdev/docker-nginx/graphs/contributors">
  <img src="https://opencollective.com/wod/contributors.svg?width=890&button=false" alt="Contributors">
</a>

<br>


CHANGELOG.md
# Changelog

## [1.5.24](https://github.com/wayofdev/docker-project-services/compare/v1.5.23...v1.5.24) (2023-11-29)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v23.0.1 ([d510352](https://github.com/wayofdev/docker-project-services/commit/d5103526ef5d17f9eca134703aa5c577bb5c79d2))

## [1.5.23](https://github.com/wayofdev/docker-project-services/compare/v1.5.22...v1.5.23) (2023-11-23)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v23 ([13c4267](https://github.com/wayofdev/docker-project-services/commit/13c426785aeefd70e1ad319a204ef2cbd8dc9aae))

## [1.5.22](https://github.com/wayofdev/docker-project-services/compare/v1.5.21...v1.5.22) (2023-10-24)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22.0.5 ([f3ccabc](https://github.com/wayofdev/docker-project-services/commit/f3ccabc2d67f2f4729ecc003f61d9ee7437303e3))

## [1.5.21](https://github.com/wayofdev/docker-project-services/compare/v1.5.20...v1.5.21) (2023-10-04)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22.0.4 ([896eed2](https://github.com/wayofdev/docker-project-services/commit/896eed282f98cff7f3f8720d39c5163cad295bf7))

## [1.5.20](https://github.com/wayofdev/docker-project-services/compare/v1.5.19...v1.5.20) (2023-10-03)


### Miscellaneous

* **deps:** update redhat-plumbers-in-action/differential-shellcheck action to v5 ([ee2f55b](https://github.com/wayofdev/docker-project-services/commit/ee2f55b7aa84d7fbee2957685a2106cb8c910373))

## [1.5.19](https://github.com/wayofdev/docker-project-services/compare/v1.5.18...v1.5.19) (2023-09-24)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1.1.8 ([01dffa6](https://github.com/wayofdev/docker-project-services/commit/01dffa6db15e93d2f4df379e949991c0c065c306))

## [1.5.18](https://github.com/wayofdev/docker-project-services/compare/v1.5.17...v1.5.18) (2023-09-24)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1.1.4 ([0746a3b](https://github.com/wayofdev/docker-project-services/commit/0746a3b6126369533c6e41b41b19da219c644844))

## [1.5.17](https://github.com/wayofdev/docker-project-services/compare/v1.5.16...v1.5.17) (2023-09-16)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1.1.3 ([a6b11ed](https://github.com/wayofdev/docker-project-services/commit/a6b11ed6cfc9de3330b237ce413b58a83966d667))

## [1.5.16](https://github.com/wayofdev/docker-project-services/compare/v1.5.15...v1.5.16) (2023-09-15)


### Miscellaneous

* **deps:** update postgres docker tag to v16 ([d43330e](https://github.com/wayofdev/docker-project-services/commit/d43330ec31b5cdd63fe6a4731812cd5819018dd9))

## [1.5.15](https://github.com/wayofdev/docker-project-services/compare/v1.5.14...v1.5.15) (2023-09-12)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22.0.3 ([486f40d](https://github.com/wayofdev/docker-project-services/commit/486f40d9ba09fe34ee702a5c11906a285dbc91c0))

## [1.5.14](https://github.com/wayofdev/docker-project-services/compare/v1.5.13...v1.5.14) (2023-09-12)


### Miscellaneous

* **deps:** update docker/setup-qemu-action action to v3 ([99292f3](https://github.com/wayofdev/docker-project-services/commit/99292f37f4976ac07707cae6b2e97c7e1ffde3ae))

## [1.5.13](https://github.com/wayofdev/docker-project-services/compare/v1.5.12...v1.5.13) (2023-09-12)


### Miscellaneous

* **deps:** update docker/setup-buildx-action action to v3 ([098e9ec](https://github.com/wayofdev/docker-project-services/commit/098e9ec043e2e88d97207bbe04a1c9759856dd31))

## [1.5.12](https://github.com/wayofdev/docker-project-services/compare/v1.5.11...v1.5.12) (2023-09-11)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22.0.2 ([1c36692](https://github.com/wayofdev/docker-project-services/commit/1c366925e3b74b38198cd7208b12c2a0d0835e7f))

## [1.5.11](https://github.com/wayofdev/docker-project-services/compare/v1.5.10...v1.5.11) (2023-09-04)


### Miscellaneous

* **deps:** update actions/checkout action to v4 ([ef1e333](https://github.com/wayofdev/docker-project-services/commit/ef1e333d5e9446e27681c2c965dd8ef7083a5aaa))

## [1.5.10](https://github.com/wayofdev/docker-project-services/compare/v1.5.9...v1.5.10) (2023-09-03)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1.1.2 ([acbbafc](https://github.com/wayofdev/docker-project-services/commit/acbbafc5692f8de3145d5f22d7de6b2a4270705e))

## [1.5.9](https://github.com/wayofdev/docker-project-services/compare/v1.5.8...v1.5.9) (2023-07-22)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1.1.1 ([2b0067e](https://github.com/wayofdev/docker-project-services/commit/2b0067efc41d174c0df51fcbefc771250b59ed05))

## [1.5.8](https://github.com/wayofdev/docker-project-services/compare/v1.5.7...v1.5.8) (2023-07-18)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22.0.1 ([5b23591](https://github.com/wayofdev/docker-project-services/commit/5b23591f27575eb4ce25a6f7311ff0ccb970cd29))

## [1.5.7](https://github.com/wayofdev/docker-project-services/compare/v1.5.6...v1.5.7) (2023-07-17)


### Miscellaneous

* **deps:** update andreasaugustin/actions-template-sync action to v1 ([18346da](https://github.com/wayofdev/docker-project-services/commit/18346da1222295329e7a8df9518bd2ea744951e7))

## [1.5.6](https://github.com/wayofdev/docker-project-services/compare/v1.5.5...v1.5.6) (2023-07-11)


### Miscellaneous

* **deps:** update quay.io/keycloak/keycloak docker tag to v22 ([f8d7bfe](https://github.com/wayofdev/docker-project-services/commit/f8d7bfe0673c8329509aad1f327cbc2b91569e58))

## [1.5.5](https://github.com/wayofdev/docker-project-services/compare/v1.5.4...v1.5.5) (2023-06-28)


### Miscellaneous


keycloak/setup/setup-admin-realm.sh
#!/bin/sh

set -euxv

KCADM=/opt/keycloak/bin/kcadm.sh

${KCADM} config credentials --server http://localhost:8080 --realm master --user "$KEYCLOAK_ADMIN" --password "$KEYCLOAK_ADMIN_PASSWORD"

${KCADM} create realms \
    -s "realm=${OIDC_ADMIN_REALM}" \
    -s enabled=true \
    -s 'displayName="Company Employees"' \
    -s accessTokenLifespan=300 \
    -s ssoSessionIdleTimeout=3600 \
    -s ssoSessionMaxLifespan=3600 \
    -s revokeRefreshToken=true \
    -s refreshTokenMaxReuse=0 \
    -s defaultSignatureAlgorithm=RS256

${KCADM} create clients -r ${OIDC_ADMIN_REALM} -f - <<EOF
{
    "protocol": "openid-connect",
    "clientId": "${OIDC_ADMIN_APP_ID}",
    "directAccessGrantsEnabled": true,
    "webOrigins": ["*"],
    "standardFlowEnabled": false,
    "clientAuthenticatorType": "client-secret",
    "secret": "${OIDC_ADMIN_APP_SECRET}",
    "bearerOnly": false,
    "publicClient": false,
    "serviceAccountsEnabled": true,
    "authorizationServicesEnabled": true,
    "attributes": {
        "client_credentials.use_refresh_token": true
    }
}
EOF

${KCADM} create clients -r ${OIDC_ADMIN_REALM} -f - <<EOF
{
    "protocol": "openid-connect",
    "clientId": "${OIDC_ADMIN_USER_ID}",
    "directAccessGrantsEnabled": true,
    "webOrigins": ["*"],
    "bearerOnly": false,
    "standardFlowEnabled": false,
    "publicClient": true,
    "serviceAccountsEnabled": false,
    "authorizationServicesEnabled": false,
    "attributes": {
        "client_credentials.use_refresh_token": true
    }
}
EOF

### Users
${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=john.doe -s email=john.doe@wayof.dev -s firstName=John -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username john.doe --new-password john.doe

${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=jane.doe -s email=jane.doe@wayof.dev -s firstName=Jane -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username jane.doe --new-password jane.doe

${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=johnny.doe -s email=johnny.doe@wayof.dev -s firstName=Johnny -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username johnny.doe --new-password johnny.doe

### Roles
${KCADM} add-roles \
    -r ${OIDC_ADMIN_REALM} \
    --uusername "service-account-${OIDC_ADMIN_APP_ID}" \
    --cclientid realm-management \
    --rolename manage-users


keycloak/setup/setup-operations-realm.sh
#!/bin/sh

set -euxv

KCADM=/opt/keycloak/bin/kcadm.sh

${KCADM} config credentials --server "$KC_INTERNAL_HOSTNAME_URL" --realm master --user "$KEYCLOAK_ADMIN" --password "$KEYCLOAK_ADMIN_PASSWORD"

${KCADM} create realms \
    -s "realm=${OIDC_ADMIN_REALM}" \
    -s enabled=true \
    -s displayName="${OIDC_ADMIN_REALM_DISPLAY_NAME}" \
    -s accessTokenLifespan=300 \
    -s ssoSessionIdleTimeout=3600 \
    -s ssoSessionMaxLifespan=3600 \
    -s revokeRefreshToken=true \
    -s refreshTokenMaxReuse=0 \
    -s defaultSignatureAlgorithm=RS256 \
    -s editUsernameAllowed=false \
    -s registrationEmailAsUsername=true \
    -s registrationAllowed=false \
    -s resetPasswordAllowed=true \
    -s rememberMe=true \
    -s duplicateEmailsAllowed=false \
    -s loginTheme=keywind

${KCADM} create clients -r ${OIDC_ADMIN_REALM} -f - <<EOF
{
    "protocol": "openid-connect",
    "clientId": "${OIDC_ADMIN_APP_ID}",
    "directAccessGrantsEnabled": true,
    "webOrigins": ["*"],
    "redirectUris": ["*"],
    "standardFlowEnabled": true,
    "clientAuthenticatorType": "client-secret",
    "secret": "${OIDC_ADMIN_APP_SECRET}",
    "bearerOnly": false,
    "publicClient": false,
    "serviceAccountsEnabled": true,
    "authorizationServicesEnabled": true,
    "attributes": {
        "client_credentials.use_refresh_token": true
    }
}
EOF

${KCADM} create clients -r ${OIDC_ADMIN_REALM} -f - <<EOF
{
    "protocol": "openid-connect",
    "clientId": "${OIDC_ADMIN_USER_ID}",
    "directAccessGrantsEnabled": true,
    "webOrigins": ["*"],
    "redirectUris": ["*"],
    "bearerOnly": false,
    "standardFlowEnabled": true,
    "publicClient": true,
    "serviceAccountsEnabled": false,
    "authorizationServicesEnabled": false,
    "attributes": {
        "client_credentials.use_refresh_token": true
    }
}
EOF

### Users
${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=john.doe@wayof.dev -s email=john.doe@wayof.dev -s firstName=John -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username john.doe@wayof.dev --new-password john.doe

${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=jane.doe@wayof.dev -s email=jane.doe@wayof.dev -s firstName=Jane -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username jane.doe@wayof.dev --new-password jane.doe

${KCADM} create users -r ${OIDC_ADMIN_REALM} -s enabled=true -s emailVerified=true -s username=johnny.doe@wayof.dev -s email=johnny.doe@wayof.dev -s firstName=Johnny -s lastName=Doe
${KCADM} set-password -r ${OIDC_ADMIN_REALM} --username johnny.doe@wayof.dev --new-password johnny.doe

### Roles
${KCADM} add-roles \
    -r ${OIDC_ADMIN_REALM} \
    --uusername "service-account-${OIDC_ADMIN_APP_ID}" \
    --cclientid realm-management \
    --rolename manage-users


keycloak/setup/setup-client-realm.sh
#!/bin/sh

set -euxv

KCADM=/opt/keycloak/bin/kcadm.sh

${KCADM} config credentials --server "$KC_INTERNAL_HOSTNAME_URL" --realm master --user "$KEYCLOAK_ADMIN" --password "$KEYCLOAK_ADMIN_PASSWORD"

${KCADM} create realms \
    -s "realm=${OIDC_CLIENT_REALM}" \
    -s enabled=true \
    -s displayName="${OIDC_CLIENT_REALM_DISPLAY_NAME}" \
    -s accessTokenLifespan=300 \
    -s ssoSessionIdleTimeout=3600 \
    -s ssoSessionMaxLifespan=3600 \
    -s revokeRefreshToken=true \
    -s refreshTokenMaxReuse=0 \
    -s defaultSignatureAlgorithm=RS256 \
    -s editUsernameAllowed=true \
    -s registrationEmailAsUsername=true \
    -s registrationAllowed=true \
    -s resetPasswordAllowed=true \
    -s rememberMe=true \
    -s duplicateEmailsAllowed=false \
    -s loginTheme=keywind

${KCADM} create clients -r "${OIDC_CLIENT_REALM}" -f - <<EOF
{
      "protocol": "openid-connect",
      "clientId": "${OIDC_CLIENT_APP_ID}",
      "directAccessGrantsEnabled": true,
      "webOrigins": ["*"],
      "redirectUris": ["*"],
      "standardFlowEnabled": true,
      "clientAuthenticatorType": "client-secret",
      "secret": "${OIDC_CLIENT_APP_SECRET}",
      "bearerOnly": false,
      "publicClient": false,
      "serviceAccountsEnabled": true,
      "authorizationServicesEnabled": true,
      "attributes": {
            "client_credentials.use_refresh_token": true
      }
}
EOF

${KCADM} create clients -r "${OIDC_CLIENT_REALM}" -f - <<EOF
{
    "protocol": "openid-connect",
    "clientId": "${OIDC_CLIENT_USER_ID}",
    "directAccessGrantsEnabled": true,
    "webOrigins": ["*"],
    "redirectUris": ["*"],
    "bearerOnly": false,
    "standardFlowEnabled": true,
    "publicClient": true,
    "serviceAccountsEnabled": false,
    "authorizationServicesEnabled": false,
    "attributes": {
        "client_credentials.use_refresh_token": true
    }
}
EOF

### Users
${KCADM} create users -r "${OIDC_CLIENT_REALM}" -s enabled=true -s emailVerified=true -s username=john.roe@wayof.dev -s email=john.roe@wayof.dev -s firstName=John -s lastName=Roe
${KCADM} set-password -r "${OIDC_CLIENT_REALM}" --username john.roe@wayof.dev --new-password john.roe

${KCADM} create users -r "${OIDC_CLIENT_REALM}" -s enabled=true -s emailVerified=true -s username=jane.roe@wayof.dev -s email=jane.roe@wayof.dev -s firstName=Jane -s lastName=Roe
${KCADM} set-password -r "${OIDC_CLIENT_REALM}" --username jane.roe@wayof.dev --new-password jane.roe

${KCADM} create users -r "${OIDC_CLIENT_REALM}" -s enabled=true -s emailVerified=true -s username=johnny.roe@wayof.dev -s email=johnny.roe@wayof.dev -s firstName=Johnny -s lastName=Roe
${KCADM} set-password -r "${OIDC_CLIENT_REALM}" --username johnny.roe@wayof.dev --new-password johnny.roe

### Roles
${KCADM} add-roles \
    -r "${OIDC_CLIENT_REALM}" \
    --uusername "service-account-${OIDC_CLIENT_APP_ID}" \
    --cclientid realm-management \
    --rolename manage-users

### Identity Providers
${KCADM} create identity-provider/instances -r "${OIDC_CLIENT_REALM}" -f - <<EOF
{
    "alias": "google",
    "providerId": "google",
    "enabled": true,
    "trustEmail": true,
    "config": {
        "useJwksUrl": true,
        "clientId": "${OIDC_IDP_GOOGLE_CLIENT_ID}",
        "clientSecret": "${OIDC_IDP_GOOGLE_CLIENT_SECRET}"
    }
}
EOF

${KCADM} create identity-provider/instances -r "${OIDC_CLIENT_REALM}" -f - <<EOF
{
    "alias": "facebook",
    "providerId": "google",
    "enabled": true,
    "trustEmail": true,
    "config": {
        "useJwksUrl": true,
        "clientId": "${OIDC_IDP_FACEBOOK_CLIENT_ID}",
        "clientSecret": "${OIDC_IDP_FACEBOOK_CLIENT_SECRET}"
    }
}
EOF
